Parsedown.php

```diff
@@ -1485,10 +1503,80 @@ function parse($text)
         return $markup;
     }
 
+    protected function sanitiseElement(array $Element)
+    {
+        static $badAttributeChars = "\"'= \t\n\r\0\x0B";
+        static $safeUrlNameToAtt  = array(
+            'a'   => 'href',
+            'img' => 'src',
+        );
+
+        if (isset($safeUrlNameToAtt[$Element['name']]))
+        {
+            $Element = $this->filterUnsafeUrlInAttribute($Element, $safeUrlNameToAtt[$Element['name']]);
+        }
+
+        if ( ! empty($Element['attributes']))
+        {
+            foreach ($Element['attributes'] as $att => $val)
+            {
+                # clear out nulls
+                if ($val === null)
+                {
+                    unset($Element['attributes'][$att]);
+                }
+                # filter out badly parsed attribute
+                elseif (strpbrk($att, $badAttributeChars) !== false)
+                {
+                    unset($Element['attributes'][$att]);
+                }
+            }
+
+            $onEventAttributeKeys = preg_grep('/^on/i', array_keys($Element['attributes']));
+
+            foreach ($onEventAttributeKeys as $att)
+            {
+                unset($Element['attributes'][$att]);
+            }
+        }
+
+        return $Element;
+    }
+
+    protected function filterUnsafeUrlInAttribute(array $Element, $attribute)
+    {
+        if ($this->safeLinksEnabled)
+        {
+            $safe = false;
+
+            foreach ($this->safeLinksWhitelist as $scheme)
+            {
+                if (stripos($Element['attributes'][$attribute], $scheme) === 0)
```

Comment:
Also, I'd just like to point out a great example of why blacklists are hard to get right

> > Blacklists are hard to get right
> That's not true. javascript: is the only protocol you must detect and remove to prevent XSS attacks. Everything else is a false positive.

`data:text/html,<script>alert(1)</script>` is xss

😉